src/string/is_valid_email_address.rs

```diff
@@ -0,0 +1,213 @@
+const MAX_LOCAL_PART_LENGTH: u32 = 64;
+const MAX_DOMAIN_LENGTH: u32 = 255;
+const CHARACTERS: &str = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz1234567890";
+const SYMBOLS: &str = "!#$%&'*+-/=?^_`{|}~";
+
+fn clear_quotes(input: &str) -> String {
+    let mut quote_count = input.starts_with('"') as u32;
+    let mut new_local_part: String = String::from(input.chars().next().unwrap());
+    for i in 1..input.len() {
+        if input.chars().nth(i).unwrap() == '"' && input.chars().nth(i - 1).unwrap() != '\\' {
+            quote_count += 1;
+
+            if !new_local_part.starts_with('"') && quote_count != 1 {
+                new_local_part.push('"');
+            }
+        }
+
+        if quote_count % 2 == 0 {
+            new_local_part.push(input.chars().nth(i).unwrap());
+        }
+    }
+
+    new_local_part
+}
+
+fn has_bad_dots(part: &str) -> bool {
+    if part.starts_with('.') || part.ends_with('.') || part.contains("..") {
+        return true;
+    }
+    false
+}
+
+fn contains_legal_local_characters(local_part: &str) -> bool {
+    for item in local_part.chars() {
+        if !CHARACTERS.contains(item) && !SYMBOLS.contains(item) && item != '.' && item != '"' {
+            return false;
+        }
+    }
+    true
+}
+
+fn contains_legal_domain_characters(domain: &str) -> bool {
+    for item in domain.chars() {
+        if !CHARACTERS.contains(item) && item != '-' && item != '.' && !"[]:".contains(item) {
+            return false;
+        }
+    }
+    true
+}
+
+fn has_valid_quotes(part: &str) -> bool {
+    for i in 1..part.len() {
+        if part.chars().nth(i - 1).unwrap() == '"' && part.chars().nth(i).unwrap() == '"' {
+            let proceeding_quote = part.chars().nth(i + 1);
+            if proceeding_quote.is_some() && proceeding_quote.unwrap() != '.' {
+                return false;
+            }
+        }
+    }
+    true
+}
+
+fn is_valid_ip(domain: &str) -> bool {
+    if domain.len() < 2 {
+        return true;
+    }
+
+    // check 'I' for the [IPv6]
+    if domain.starts_with('[')
+        && domain.chars().nth(1).unwrap() != 'I'
+        && !domain.chars().nth(1).unwrap().is_ascii_digit()
+    {
+        return false;
+    }
+
+    if domain.starts_with('[') && domain.chars().nth(1).unwrap().is_ascii_digit() {
+        let sections = domain[1..domain.len() - 1]
+            .split('.')
+            .map(|x| x.to_string())
+            .collect::<Vec<String>>();
+        for section in sections {
+            let value = section.parse::<u32>();
+            if value.is_err() || value.unwrap() > 255 {
+                return false;
+            }
+        }
+    }
+
+    true
+}
```

Comment:
Why not to split it into two functions: `is_ipv4` and `is_ipv6`? This will allow for testing them directly (currently lots of testes are missing e.g. `123.123.123.123.123.123.123.123` is not a valid ip address).

Furthermore, the logic here is quite convoluted. What do you think about checking (in `is_domain_valid`) if the _domain part_ has the form `[<some_str>]`, and if so, send the `<some_str>` to `is_ip` function. It could look like
```rust
fn is_valid_ip(ip_str: &str) -> bool {
    is_ipv4(ip_str) || is_ipv6(ip_str)
}
```

Such approach would me more natural: what is a valid ip address `255.255.255.255` or `[255.255.255.255]`?